File: python/binary_search/find_first_and_last_range.py
# Given an array of integers nums sorted in non-decreasing order, find the starting and ending position of a given target value.
# If target is not found in the array, return [-1, -1].
# You must write an algorithm with O(log n) runtime complexity.

class Solution:
    def searchRange(self, nums: list[int], target: int) -> list[int]:
        if not nums:
            return [-1, -1]

        left = self.find_left(nums, target)
        if left == -1:
            return [-1, -1]

        right = self.find_right(nums, target)

        return [left, right]

    def find_left(self, nums: list[int], target: int) -> int:
        left, right = 0, len(nums) - 1
        while left <= right:
            mid = (left + right) // 2
            if nums[mid] >= target:
                right = mid - 1
            else:
                left = mid + 1

        # left stops at the first position where nums[left] >= target or len(nums) if not found
        if left < len(nums) and nums[left] == target:
            return left
        return -1

    def find_right(self, nums: list[int], target: int) -> int:
        left, right = 0, len(nums) - 1
        while left <= right:
            mid = (left + right) // 2
            if nums[mid] <= target:
                left = mid + 1
            else:
                right = mid - 1

        # right stops at the last position where nums[right] <= target or -1 if not found
        if right >= 0 and nums[right] == target:
            return right
        return -1

File: java/Hot100/BinarySearch/FindFirstAndLastRange.java
package Hot100.BinarySearch;

/**
 * Given an array of integers nums sorted in non-decreasing order, find the starting and ending position of a given target value.
 * <p>
 * If target is not found in the array, return [-1, -1].
 * <p>
 * You must write an algorithm with O(log n) runtime complexity.
 */
public class FindFirstAndLastRange {
    public int[] searchRange(int[] nums, int target) {
        if(nums.length==0){
            return new int[]{-1,-1};
        }

        int left = findLeft(nums, target);
        if (left == -1){
            return new int[]{-1, -1};
        }
        int right = findRight(nums, target);
        return new int[]{left, right};
    }

    private int findLeft(int[] nums, int target) {
        int left = 0, right = nums.length - 1;
        while (left < right) {
            int mid = (left + right) / 2;
            if (nums[mid] >= target) {
                right = mid;  // keep the mid because nums[mid] might equal the target
            } else {
                left = mid + 1;
            }
        }
        return nums[left] == target ? left : -1;
    }

    private int findRight(int[] nums, int target) {
        int left = 0, right = nums.length - 1;
        while (left < right) {
            int mid = (left + right + 1) / 2; // plus 1 to ensure moving right
            if (nums[mid] <= target) {
                left = mid;  // // keep the mid because nums[mid] might equal the target
            } else {
                right = mid - 1;
            }
        }
        return nums[left] == target ? left : -1;
    }
}
